src/Servers/Kestrel/Core/src/Middleware/TlsListenerMiddleware.cs

```diff
@@ -0,0 +1,125 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using Microsoft.AspNetCore.Connections;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Middleware;
+
+internal sealed class TlsListenerMiddleware
+{
+    private readonly ConnectionDelegate _next;
+    private readonly Action<ConnectionContext, ReadOnlySequence<byte>> _tlsClientHelloBytesCallback;
+
+    public TlsListenerMiddleware(ConnectionDelegate next, Action<ConnectionContext, ReadOnlySequence<byte>> tlsClientHelloBytesCallback)
+    {
+        _next = next;
+        _tlsClientHelloBytesCallback = tlsClientHelloBytesCallback;
+    }
+
+    /// <summary>
+    /// Sniffs the TLS Client Hello message, and invokes a callback if found.
+    /// </summary>
+    internal async Task OnTlsClientHelloAsync(ConnectionContext connection)
+    {
+        var input = connection.Transport.Input;
+
+        while (true)
+        {
+            var result = await input.ReadAsync();
+            var buffer = result.Buffer;
+
+            // If the buffer length is less than 6 bytes (handshake + version + length + client-hello byte)
+            // and no more data is coming, we can't block in a loop here because we will not get more data
+            if (buffer.Length < 6 && result.IsCompleted)
+            {
+                break;
+            }
+
+            var parseState = TryParseClientHello(buffer, out var clientHelloBytes);
+
+            // no data is consumed, it will be processed by the follow-up middlewares
+            input.AdvanceTo(buffer.Start);
+
+            switch (parseState)
+            {
+                case ClientHelloParseState.NotEnoughData:
+                    continue;
+
+                case ClientHelloParseState.NotTlsClientHello:
+                    await _next(connection);
+                    return;
+
+                case ClientHelloParseState.ValidTlsClientHello:
+                    _tlsClientHelloBytesCallback(connection, clientHelloBytes);
+                    await _next(connection);
+                    return;
+            }
```

Comment:
To reduce the IL size of this method by reducing the async statemachinery the switch could be written as
```suggestion
            switch (parseState)
            {
                case ClientHelloParseState.NotEnoughData:
                    continue;

                case ClientHelloParseState.ValidTlsClientHello:
                    _tlsClientHelloBytesCallback(connection, clientHelloBytes);
                    goto case default;

                default:
                    await _next(connection);
                    return;
            }
```
So from `case ClientHelloParseState.ValidTlsClientHello` it falls through to the `default` case, thus only 1 async state machinerey is needed to be emitted by C# compiler (instead of two).